build/templates/mock_helper.py.mako

```diff
@@ -65,7 +65,8 @@ ivi_dance_param = helper.extract_ivi_dance_parameter(params)
 ivi_dance_size_param = helper.find_size_parameter(ivi_dance_param, params)
 %>\
     def ${c_function_prefix}${func_name}(${helper.get_params_snippet(f, helper.ParamListType.LIBRARY_IMPL_METHOD)}):  # noqa: N802
-        if self._defaults['${func_name}']['return'] != 0:
+        if self._defaults['${func_name}']['return'] < 0:
```

Comment:
This means the string value will be set even during the first part of the IVI dance.

I think maybe the correct way to fix this would be to insert the original if statement, but only when the function supports the ivi dance style of string retrieval. In other cases, it would fall through to the normal return.

Reply:
> This means the string value will be set even during the first part of the IVI dance.

That's right.

> I think maybe the correct way to fix this would be to insert the original if statement, but only when the function supports the ivi dance style of string retrieval. In other cases, it would fall through to the normal return.

I think that's getting too clever. Unit test code should have very few logic and no codepaths. The caller should explicitly program this behavior. Can't we set expectations for first and second call separately?

Reply:
But then a test could be written that works because of the incorrect behavior of the mock, but wouldn't work on the real driver.

I think the mock should mock the behavior of the driver as closely as possible.

What do drivers do for other functions that return a value when they get an error? Looking at scope, it looks like it doesn't return data if there is an error. I was looking at some of the `CalFetch*()` functions.

Reply:
> But then a test could be written that works because of the incorrect behavior of the mock, but wouldn't work on the real driver.

This is a reality of unit testing. If the developer programs the mock to behave differently than the object it's mocking, then the code won't be properly tested.

> I think the mock should mock the behavior of the driver as closely as possible.

This is true. Where that responsibility falls is the question.

> What do drivers do for other functions that return a value when they get an error? Looking at scope, it looks like it doesn't return data if there is an error. I was looking at some of the CalFetch*() functions.

It varies. For example in the fetch and timeout case: NI-DMM and NI-DCPower will return the points it managed to get before the timeout.

Reply:
> > But then a test could be written that works because of the incorrect behavior of the mock, but wouldn't work on the real driver.

> This is a reality of unit testing. If the developer programs the mock to behave differently than the object it's mocking, then the code won't be properly tested.

Then why are we programming the mock to behave differently?

> > I think the mock should mock the behavior of the driver as closely as possible.

> This is true. Where that responsibility falls is the question.

I think the responsibility belongs here. And if that adds complexity to the generator so that the generated side effect function is "correct", then I think that is the appropriate tradeoff.

An alternative would be to not use the generated side effect function and instead write one by hand in the test file that has the appropriate behavior.

Reply:
I think the default side effect should stay as it is, since that is what is needed most of the time.

When you need different behavior, like you do here, then I think the correct answer is to create a custom side effect function at the place where you need to different behavior. You can see where I did this in the ModInst unit tests. I needed a side effect function for testing iterating functionality.